src/Pos/Block/Network/Retrieval.hs

```diff
@@ -76,25 +56,15 @@ retrievalWorker = worker outs retrievalWorkerImpl
                              (Proxy :: Proxy (MsgBlock ssc))
                       ]
 
-needRecovery :: forall ssc m.
-       (SscWorkersClass ssc, WorkMode ssc m)
-       => Proxy ssc -> m Bool
-needRecovery _ = getCurrentSlot >>= maybe (pure True) needRecoveryCheck
-  where
-    needRecoveryCheck slot = do
-        header <- DB.getTipBlockHeader @ssc
-        return $ toEnum (fromEnum (getEpochOrSlot header) + blkSecurityParam) < getEpochOrSlot slot
-
 retrievalWorkerImpl
     :: forall ssc m.
        (SscWorkersClass ssc, WorkMode ssc m)
     => SendActions m -> m ()
 retrievalWorkerImpl sendActions = handleAll handleTop $ do
     logDebug "Starting retrievalWorker loop"
-    NodeContext{..} <- getNodeContext
-    mainLoop ncBlockRetrievalQueue ncRecoveryHeader
-  where
-    mainLoop queue recHeaderVar = runIfNotShutdown $ reportingFatal version $ do
+    forever $ runIfNotShutdown $ reportingFatal version $ do
```

Comment:
@georgeee no, in this case check inside `runIfNotShutdown` would be performed only once, and it won't work at all.
In case of `forever . runIfNotShutdown` it probably would work, but there would be a ton of shutdown notifications spawned to the queue, which is bad too.
I would vote for returning to `mainLoop` pattern.